bench/store_bench_test.go

```diff
@@ -0,0 +1,248 @@
+}
+
+func BenchmarkStore_MemoryPut_W1(b *testing.B) {
+	benchmarkStorePut(b, newMemorySubject, workload1(b))
+}
+
+func BenchmarkStore_MemoryGet_W1(b *testing.B) {
+	benchmarkStoreGet(b, newMemorySubject, workload1(b))
+}
+
+func BenchmarkStore_MemoryPut_W2(b *testing.B) {
+	benchmarkStorePut(b, newMemorySubject, workload2(b))
+}
+
+func BenchmarkStore_MemoryGet_W2(b *testing.B) {
+	benchmarkStoreGet(b, newMemorySubject, workload2(b))
+}
+
+func BenchmarkStore_MemoryPut_W3(b *testing.B) {
+	benchmarkStorePut(b, newMemorySubject, workload3(b))
+}
+
+func BenchmarkStore_MemoryGet_W3(b *testing.B) {
+	benchmarkStoreGet(b, newMemorySubject, workload2(b))
+}
+
+func newPebbleSubject(b *testing.B) func() (indexer.Interface, error) {
+	return func() (indexer.Interface, error) {
+		// Default options copied from cockroachdb with the addition of 1GiB cache.
+		// See:
+		// - https://github.com/cockroachdb/cockroach/blob/v22.1.6/pkg/storage/pebble.go#L479
+		pebbleOpts := &pb2.Options{
+			BytesPerSync:                10 << 20, // 10 MiB
+			WALBytesPerSync:             10 << 20, // 10 MiB
+			MaxConcurrentCompactions:    10,
+			MemTableSize:                64 << 20, // 64 MiB
+			MemTableStopWritesThreshold: 4,
+			LBaseMaxBytes:               64 << 20, // 64 MiB
+			L0CompactionThreshold:       2,
+			L0StopWritesThreshold:       1000,
+			WALMinSyncInterval:          func() time.Duration { return 30 * time.Second },
+		}
+
+		pebbleOpts.Experimental.ReadCompactionRate = 10 << 20 // 20 MiB
+		pebbleOpts.Experimental.MinDeletionRate = 128 << 20   // 128 MiB
+
+		const numLevels = 7
+		pebbleOpts.Levels = make([]pb2.LevelOptions, numLevels)
+		for i := 0; i < numLevels; i++ {
+			l := &pebbleOpts.Levels[i]
+			l.BlockSize = 32 << 10       // 32 KiB
+			l.IndexBlockSize = 256 << 10 // 256 KiB
+			l.FilterPolicy = bloom.FilterPolicy(10)
+			l.FilterType = pb2.TableFilter
+			if i > 0 {
+				l.TargetFileSize = pebbleOpts.Levels[i-1].TargetFileSize * 2
+			}
+			l.EnsureDefaults()
+		}
+		pebbleOpts.Levels[numLevels-1].FilterPolicy = nil
+		pebbleOpts.Cache = pb2.NewCache(1 << 30) // 1 GiB
+
+		return pebble.New(b.TempDir(), pebbleOpts)
+	}
+}
+
+func sthSubject(b *testing.B) func() (indexer.Interface, error) {
+	return func() (indexer.Interface, error) {
+		return storethehash.New(context.Background(), b.TempDir(),
+			indexer.BinaryValueCodec{},
+			64,
+			sth.GCInterval(30*time.Minute),
+			sth.GCTimeLimit(5*time.Minute),
+			sth.SyncInterval(3*time.Second),
+			sth.IndexBitSize(24),
+		)
+	}
+}
+
+func newMemorySubject() (indexer.Interface, error) {
+	return memory.New(), nil
+}
+
+func benchmarkStorePut(b *testing.B, newSubject func() (indexer.Interface, error), w *workload) {
+	subject, err := newSubject()
+	if err != nil {
+		b.Fatal(err)
+	}
+	defer func() {
+		if err := subject.Close(); err != nil {
+			b.Fatal()
+		}
+	}()
+	b.SetBytes(int64(w.size))
+	b.ReportAllocs()
+	b.ResetTimer()
+	b.RunParallel(func(pb *testing.PB) {
```

Comment:
I am not sure this is what we want here. My understanding is that `RunParallel` is for doing CPU benchmarking, so that you distribute portions of the test executions over your cores in order to make the CPU maximally busy with your test iterations.

In the case of the benchmarks here, the code we are benchmarking handles concurrency. So, I think we want to remove `RunParallel` here in order to test our code's ability to use concurrency.

Reply:
Using `RunParallel` effectively benchmarks contention of inserting lots of values at the same time, where puts are done across multiple coroutines. and each does a sequential iteration through the entire list of values to insert it. No?

Reply:
In any case, here are the benchmark results on my machine running without parallelism using 
```
b.Run("_", func(b *testing.B) {
		for i := 0; i < b.N; i++ { ...
```

```
name                                time/op
Store_Sth_PutConcurrency_W0_1/_-8   930ms ± 1%
Store_Sth_PutConcurrency_W0_2/_-8   1.66s ±10%
Store_Sth_PutConcurrency_W0_64/_-8  1.63s ± 4%
```